File: 907-koko-eating-bananas/koko-eating-bananas.cpp
class Solution
{
public:
    int minEatingSpeed(vector<int> &piles, int h)
    {
        int l = 1, r = 1e9;
        while (l < r)
        {
            int m = (l + r) / 2, t = 0;
            for (int p : piles)
                t += (p + m - 1) / m;
            if (t > h)
                l = m + 1;
            else
                r = m;
        }
        return l;
    }
};

File: 907-koko-eating-bananas/koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left , right = 1 , max(piles)
        res = right

        while(left <= right):
            mid = (left + right) // 2
            hours = 0
            for n in piles:
                hours += (n + mid - 1) // mid
            
            if hours <= h:
                res = mid
                right = mid - 1
            else:
                left = mid + 1
        
        return res

